Add type annotations to gdb/__init__.py

slackito opened the pull request.

slackito committed:
Add type annotations to gdb/__init__.py

rupprecht commented on gdb/__init__.py:
```diff
@@ -398,19 +408,19 @@ def _get_child_member_with_name(sbvalue, name):
         return self._sbvalue_object
 
-    def _stripped_sbtype(self):
+    def _stripped_sbtype(self) -> tuple[lldb.SBType, int]:
```
tuple -> typing.Tuple, for older python3 compatibility

slackito replied:
Done.

rupprecht commented on gdb/__init__.py:
```diff
@@ -317,31 +326,31 @@ def target(self):
         return Type(self._sbtype_object.GetPointerType())
 
-    def template_argument(self, n):
+    def template_argument(self, n: int) -> 'Type | Value':
```
`|` appears to be only in 3.10, so this should be `Union['Type', 'Value']` for now

slackito replied:
Done.

rupprecht commented on gdb/__init__.py:
```diff
@@ -936,21 +935,29 @@ def lookup_type(name, block=None):
+    pass
+
+
+def current_objfile() -> Optional[Objfile]:
```
This is actually supposed to be Optional[GdbObjectFile], although GdbObjectFile is just defined to Any. I think the `class Objfile` is misleading unless I'm missing how it's used elsewhere.

rupprecht replied:
Maybe just `-> None`?

slackito replied:
According to https://sourceware.org/gdb/onlinedocs/gdb/Objfiles-In-Python.html

> When auto-loading a Python script (see [Python Auto-loading](https://sourceware.org/gdb/onlinedocs/gdb/Python-Auto_002dloading.html#Python-Auto_002dloading)), GDB sets the “current objfile” to the corresponding objfile. This function returns the current objfile. If there is no current objfile, this function returns None.

and also:

> Each objfile is represented by an instance of the gdb.Objfile class.

The `class Objfile` in this source file is GALA's implementation of `gdb.Objfile`. And `current_objfile()` is supposed to return an object of that class, or `None`. Right now we always return `None` because this feature is not implemented, but we could add it to the autoload module.

rupprecht replied:
Ah, makes sense. Can you add a brief comment here?

I had the same question about progspace, and found https://sourceware.org/gdb/onlinedocs/gdb/Progspaces-In-Python.html

slackito replied:
Done.

slackito committed:
Make type annotations compatible with older Python versions.

- Replace `X | Y` with `typing.Union[X, Y]`
- Replace `tuple[X]` with `typing.Tuple[X]`
- Also add some comments to clarify what Objfile and Progspace classes
  are for.
